### PR TITLE
Replace float16 with at::Half in caffe2

### PR DESCRIPTION
**Pull request by li-roy**
- Finishes unifying Half type in pytorch and caffe2
- As a side effect, aten_op works for fp16 now

### CODE REVIEW

File: aten/src/ATen/core/typeid.h [hunk 1]
```diff
@@ -30,6 +30,13 @@
 // fix all of the call-sites, which I want to do later.  So the namespace
 // is not fixed at the moment.
 
+// Make __f16 a fundamental type.
+namespace std {
+template<>
+struct is_fundamental<at::Half> : std::integral_constant<bool, true> {
```
Comment (ezyang) [inline aten/src/ATen/core/typeid.h]: I am not sure this is defined behavior, but I guess since Caffe2 has been doing it for a while, should be OK.

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
Please get a caffe2 person to review.

File: aten/src/ATen/core/typeid.h (hunk 1 above)
Comment (goldsborough) [reply aten/src/ATen/core/typeid.h]: Specializing a template in namespace `std` for a user-defined type is fine. `std::integral_constant<bool, true>` here should just be `std::true_type`

File: caffe2/core/types.h [hunk 2]
```diff
@@ -50,7 +51,7 @@ CAFFE2_API const TypeMeta& DataTypeToTypeMeta(const TensorProto::DataType& dt);
 static_assert(sizeof(unsigned short) == 2,
               "Short on this platform is not 16 bit.");
 namespace caffe2 {
-typedef struct CAFFE2_API CAFFE2_ALIGNED(2) __f16 { uint16_t x; } float16;
+typedef at::Half float16;
```
Comment (goldsborough) [inline caffe2/core/types.h]: Could change to `using` while we're at it

**Review by Yangqing — COMMENTED** [review]
Thanks for the PR! This looks great, own general comment about unification.

File: caffe2/core/types.h [hunk 3]
```diff
@@ -50,7 +51,7 @@ CAFFE2_API const TypeMeta& DataTypeToTypeMeta(const TensorProto::DataType& dt);
 static_assert(sizeof(unsigned short) == 2,
               "Short on this platform is not 16 bit.");
 namespace caffe2 {
-typedef struct CAFFE2_API CAFFE2_ALIGNED(2) __f16 { uint16_t x; } float16;
+using float16 = at::Half;
```
[the remark this answers is not in the document]
Comment (li-roy) [reply caffe2/core/types.h]: Managed to do this without changing too much. The main noteworthy change is the macro in `caffe2/perfkernels/embedding_lookup.cc` and codegen in caffe2/perfkernels/hp_emblookup_codegen.py.

**Comment by li-roy** [comment]
Closing in favor of #11785.

**Pull request closed without merging**